Renaming `CoreContextHandler` to `CoreAppContext` and associated files/dirs/modules.

joakime opened the pull request:
* Renaming class `CoreContextHandler` to `CoreAppContext`
* Renaming maven module `jetty-core-webapp` to `jetty-coreapp`
* Renaming jetty module `core-deploy` to `coreapp-deploy`
* Moving `jetty-core/jetty-core-test-webapps` to `jetty-core/jetty-tests/` tree
* Renaming new test coreapps to pattern `jetty-test-coreapp-<purpose>`

Closes #13237

joakime committed:
Renaming CoreContextHandler to CoreAppContext and associated files/dirs/modules.

joakime committed:
Fixing build

joakime committed:
Renameing core-deploy to coreapp-deploy

joakime committed:
Fixing lib name

joakime committed:
Fixing missed rename

janbartel commented on jetty-core/jetty-coreapp/src/main/java/org/eclipse/jetty/core/app/StaticContextHandler.java:
```diff
@@ -11,7 +11,7 @@
 //
 
-package org.eclipse.jetty.core.webapp;
+package org.eclipse.jetty.core.app;
```
I think this whole class should move back into `jetty-server` project because it  isn't a core app and doesn't depend on any other classes in this package. IMHO the only way this could stay in the `jetty-coreapp` project is if `StaticContextHandler` inherited off `CoreAppContext` (probably with a name change to `StaticContext`), and was a degenerate case with no support for `/classes, /lib, jetty-web.xml` etc etc, but that would mean that all the static content would have to be in the `/static` subdirectory of the zip.

sbordet replied:
@janbartel while you're right that `StaticContextHandler` may live in `jetty-server`, it would be weird to have its Jetty module `static-deploy.mod` live here.

If also `static-deploy.mod` is moved to `jetty-server`, then it will be a module living in `jetty-server` that however depends on a Jetty module living in `jetty-deploy` (and same with the associated Jetty module XML).

I'm fine any way, but keeping class, `*.mod` and `*.xml` together seems best, and `jetty-core-app` seems also the best place.

An embedded application would need to depend on `jetty-core-app` rather than just `jetty-server`, which is not that terrible.

In any case, I'm ok with any solution:

1. Keep everything in `jetty-core-app`
2. Move everthing to `jetty-server`
3. Split: class in `jetty-server`, modules-related files in `jetty-core-app`.

janbartel replied:
@simone don't forget that `StaticContextHandler` class doesn't depend on anything in the `jetty-deploy` module, just classes in `jetty-server` and `jetty-util` etc etc.  Only the `static-deploy.mod` file depends on the `deployment-scanner` module. I'm fine if we want to move the `StaticContextHandler` and associated module out into it's own subproject as a sibling of `jetty-coreapp`.

sbordet replied:
@janbartel I understand you like either option 2 or option 3, but not 1?

I think an own module just for `StaticContextHandler` is overkill, and like better option 3 instead of own module.

gregw replied:
The way I think about it is that `StaticContextHandler` is a component of core and is-not-a coreapp itself.   Specifically, unlike webapps, a static only core deployment is not a degenerate case of a coreapp, but is a different usage of core components to make a different type of context that can be deployed.

So it probably is justified to have its own module as a sibling to core/app

joakime replied:
jetty-server has no dependency on jetty-deploy, in fact it's the other way around (jetty-deploy depends on jetty-server).
We simply cannot test StaticContextHandler in jetty-server.

What happens in each proposed solution ...

1. Nothing changes from this PR.

2. StaticContextHandler moves to jetty-server/src/main/java
    - etc/jetty-static-deploy.xml moves to jetty-deploy/src/main/config/etc/
    - modules/static-deploy.mod moves to jetty-deploy/src/main/config/modules/
    - StaticContextHandlerTest moves to jetty-deploy/src/test/java
    - In short it doesn't just go to jetty-server, it cannot, as jetty-server doesn't have a dependency on jetty-deploy

3. StaticContextHandler moves to jetty-server/src/main/java
    - etc/jetty-static-deploy.xml stays in jetty-coreapp/src/main/config/etc/
    - modules/static-deploy.mod stays in jetty-coreapp/src/main/config/modules/
    - StaticContextHandlerTest stays in jetty-coreapp/src/test/java
    - In short it doesn't just go to jetty-server, it cannot, as jetty-server doesn't have a dependency on jetty-deploy

There is also a 4th option, move it all to jetty-deploy, resulting in ...

 - jetty-deploy/src/main/java/org/eclipse/jetty/deploy/StaticContext.java
 - jetty-deploy/src/main/config/etc/jetty-static-deploy.xml
 - jetty-deploy/src/main/config/modules/static-deploy.mod
 - jetty-deploy/src/test/java/org/eclipse/jetty/deploy/StaticContextTest.java

There's really NO reason for it to be in jetty-server, as the functionality StaticContextHandler provides is really deploy specific, and if you wanted something like it in embedded, there are easier (and more capable) setups than using StaticContextHandler.

joakime replied:
> I think this whole class should move back into `jetty-server` project because it isn't a core app and doesn't depend on any other classes in this package. 

But the testing of the `StaticContextHandler` cannot be done in jetty-server as there is no dependency on jetty-deploy there.

> IMHO the only way this could stay in the `jetty-coreapp` project is if `StaticContextHandler` inherited off `CoreAppContext` (probably with a name change to `StaticContext`), and was a degenerate case with no support for `/classes, /lib, jetty-web.xml` etc etc, but that would mean that all the static content would have to be in the `/static` subdirectory of the zip.

This was discussed with Simone and Greg a while back. (when the static-deploy was introduced)
The idea of a directory being a static deployment of it's own is desired, and is backwards compatible with previous versions of Jetty.
Where `$JETTY_BASE/webapps/name/` is all you need, the `name` is the context-path, and the contents are statically served.   In past versions of Jetty, there's no other configuration possible (eg: welcome files, cache control, dir-allowed, compressed content config, etc)
We created the `static-deploy` to allow this kind of setup still.

This cannot be done with a `CoreAppContext`, as it's an app structure (like you pointed out).
So making StaticContextHandler extend from CoreAppContext isn't a solution.

janbartel replied:
My very strong preference is that `StaticContextHandler` and associated `.mod` files move into it's own project as a sibling to `jetty-coreapp`. It is confusing and does not make sense to put it into the `jetty-coreapp` project as it is explicitly _not_ a core app.

gregw replied:
I agree with Jan.  
I think any attempt to put static mod files or code in core-app just dilutes the clarity of what a core-app is.
Splitting the static mod and class between two modules just feels unnecessary confusing.
A little simple module that represents our static content app feels clean, simple and not confusing

It is not like we don't already have infinite modules as per the 0, 1, infinity counting!

joakime replied:
What about option 4?  Put the StaticContextHandler, associated xml and mods, and tests into jetty-deploy?

joakime replied:
> My very strong preference is that `StaticContextHandler` and associated `.mod` files move into it's own project as a sibling to `jetty-coreapp`

This statement is confusing.
You go on and state that it's not a coreapp, then want it to be a sibling to coreapp, which means it's a coreapp one off.

gregw replied:
@joakime a sibling of coreapp does not mean it is a coreapp.  It means that it is something that is not a core-app.    In fact it is a static app that uses core components. 

To me, having it as a sibling makes core-app an even stronger concept.  A core-app is more than just and app made from core components, it is a very specific style of app that we have designed.

I would very much have liked for our static app to be the degenerate case of our core-app (just as a static folder can be the degenerate case of a webapp with no WEB-INF), but our design of core-app precludes that.  Trying to fit a static app inside core-app, when it is not, is just very confusing.

I have a "very strong preference" for Jan's suggestion.

joakime replied:
OK, a standalone Maven Module it is.

What do we call it?

joakime committed:
Make static-deploy its own maven module at same level as core-deploy

joakime commented:
static-deploy is now it's own maven module.

Of note, the java package name `org.eclipse.jetty.core.static` is invalid as `static` is a reserved word on Java.
So I changed it to `org.eclipse.jetty.staticapp` instead. (also changed `org.eclipse.jetty.core.app` to `org.eclipse.jetty.coreapp` to be consistent with naming)

sbordet commented on documentation/jetty/modules/operations-guide/pages/deploy/index.adoc:
```diff
@@ -27,7 +27,7 @@ This means that you can simultaneously deploy an old Java EE 8 web application,
 
 * For Jetty Static, xref:modules/standard.adoc#static-deploy[`static-deploy`]
-* For Jetty Core, xref:modules/standard.adoc#core-deploy[`core-deploy`]
+* For Jetty Core, xref:modules/standard.adoc#coreapp-deploy[`coreapp-deploy`]
```
I'd revert this, given that the `eeN-deploy` modules are not called `eeNapp-deploy`.

Also the `static-deploy` module has not been renamed to `staticapp-deploy`, so I'd go back to `core-deploy`.

sbordet commented on jetty-core/jetty-bom/pom.xml:
```diff
@@ -52,7 +52,7 @@
       <dependency>
         <groupId>org.eclipse.jetty</groupId>
-        <artifactId>jetty-core-webapp</artifactId>
+        <artifactId>jetty-coreapp</artifactId>
```
Is the new `jetty-staticapp` module also in this BOM?

joakime replied:
Yes - https://github.com/jetty/jetty.project/pull/13241/commits/60246241ea4d26d47e90382ba660cd923bf72cbe

joakime committed:
Adding jetty-staticapp to jetty-bom & jetty-home

janbartel reviewed (changes requested):
I'm with @sbordet on this, it should be `core-deploy` and `static-deploy`.

joakime committed:
coreapp-deploy to core-deploy

gregw commented on jetty-core/jetty-staticapp/pom.xml:
```diff
@@ -0,0 +1,44 @@
+  </parent>
+  <artifactId>jetty-staticapp</artifactId>
+  <name>Core :: Static App</name>
+  <description>Jetty Static App</description>
```
```suggestion
  <description>Jetty Static Content App</description>
```

gregw commented on jetty-core/jetty-coreapp/src/main/java/org/eclipse/jetty/coreapp/CoreAppContext.java:
```diff
@@ -66,9 +66,9 @@
  * </p>
  */
-public class CoreContextHandler extends ContextHandler implements Deployable
+public class CoreAppContext extends ContextHandler implements Deployable
```
I think the javadoc could be tweaked a little bit.  Something like:
```
/**
 * A specialised ContextHandler that deploys a Web Application based on Jetty core APIs.
```

I then think all the text about `metadata` is kind of meaning less without a bit more of an introduction as to the layout of a coreapp

gregw commented on jetty-core/jetty-staticapp/src/main/java/org/eclipse/jetty/staticapp/StaticAppContext.java:
```diff
@@ -40,35 +40,35 @@
  * </p>
  */
-public class StaticContextHandler extends ContextHandler
+public class StaticAppContext extends ContextHandler
```
The javadoc here could also be improved: A specialised ContextHandler that provides a Web Application that serves only static content using Jetty core components

gregw reviewed (approved):
I think the javadoc needs a bit of work, and I assume there is also documentation work to be done elsewhere.   So I'm approving this PR, so long as it is immediately followed by one on the documentation
